pkg/apis/core/v1alpha1/types_federatedobject.go

```diff
@@ -0,0 +1,254 @@
+
+	// Spec defines the desired behavior of the FederatedObject.
+	Spec GenericFederatedObjectSpec `json:"spec"`
+
+	// Status describes the most recently observed status of the FederatedObject.
+	// +optional
+	Status GenericFederatedObjectStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+
+// FederatedObjectList contains a list of FederatedObject.
+type FederatedObjectList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []FederatedObject `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:validation:Required
+// +kubebuilder:resource:path=clusterfederatedobjects,shortName=cfo,singular=clusterfederatedobject,scope=Cluster
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// ClusterFederatedObject describes a cluster-scoped Kubernetes object and how it should be propagated to different
+// member clusters.
+type ClusterFederatedObject struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired behavior of the ClusterFederatedObject.
+	Spec GenericFederatedObjectSpec `json:"spec"`
+
+	// Status describes the most recently observed status of the ClusterFederatedObject.
+	// +optional
+	Status GenericFederatedObjectStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterFederatedObjectList contains a list of ClusterFederatedObject.
+type ClusterFederatedObjectList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ClusterFederatedObject `json:"items"`
+}
+
+// GenericFederatedObjectSpec defines the desired behavior of a FederatedObject or ClusterFederatedObject.
+type GenericFederatedObjectSpec struct {
+	// Template is the base template of the Kubernetes object to be propagated.
+	Template apiextensionsv1.JSON `json:"template"`
+	// Overrides describe the overrides that should be applied to the base template of the Kubernetes object before it
+	// is propagated to individual member clusters.
+	Overrides []OverrideWithController `json:"overrides"`
+	// Placements describe the member clusters that the Kubernetes object will be propagated to, which is a union of all
+	// the listed clusters.
+	Placements []PlacementWithController `json:"placements"`
+	// Follows defines other objects, or "leaders", that the Kubernetes object should follow during propagation. I.e.
+	// the Kubernetes object should be propagated to all member clusters that its "leaders" are placed in.
+	Follows []LeaderReference `json:"follows"`
```

Comment:
Should these be `+optional` and `omitempty`?

Reply:
i thought that making these fields non-optional would remove the ambiguity (if any) when these fields are empty